### PR TITLE
ddl: improve MySQL compatibility on time-to-year conversion

### LINKED ISSUE #25331: conversion from time to year should failed
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
set time_zone = UTC;
drop table if exists tbl_2;
create table tbl_2 (col_14 time);
insert into tbl_2 values ('08:45:47.00');
alter table tbl_2 change column col_14 col_41 year;
```

### 2. What did you expect to see? (Required)

```sql
ERROR 1264 (22003): Out of range value for column 'col_41' at row 1
```

### 3. What did you see instead (Required)

```sql
Query OK, 0 rows affected (0.54 sec)
mysql> select * from tbl_2;
+--------+
| col_41 |
+--------+
|   2021 |
+--------+
```

### 4. What is your TiDB version? (Required)

master

```
commit 6d69d13d10138b8f1eab1e36a669a43c390c714e (HEAD, upstream/master)
Author: disksing <i@disksing.com>
Date:   Wed Jun 9 14:02:28 2021 +0800

    store/tikv: cleanup dependency (#25283)
```

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #25331 

Problem Summary: The cast function during changing column type should be disallowed.

### What is changed and how it works?

What's Changed: disallow changing column type from time to year

### Check List 

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
Improve MySQL compatibility on time-to-year conversion.
```

### CODE REVIEW

**Review by wjhuang2016 — COMMENTED** [review]
What does `disallow` mean?

**Comment by tangenta** [comment]
@wjhuang2016 Disallow means "forbid":
```sql
drop table if exists tbl_2;
create table tbl_2 (col_14 time);
insert into tbl_2 values ('08:45:47.00');
alter table tbl_2 change column col_14 col_41 year;
```
```console
ERROR 1264 (22003): Out of range value for column 'col_41' at row 1
```

**Comment by wjhuang2016** [comment]
> ```sql
> alter table tbl_2 change column col_14 col_41 year;
> ```

No, you are a liar :)
```
mysql> create table tbl_2 (col_14 time);
Query OK, 0 rows affected (0.01 sec)

mysql> alter table tbl_2 change column col_14 col_41 year;
Query OK, 0 rows affected (2.52 sec)
```

**Comment by tangenta** [comment]
@wjhuang2016 Maybe I should add a precondition description "non-empty".

**Comment by wjhuang2016** [comment]
> @wjhuang2016 Maybe I should add a precondition description "non-empty".

No,
```
mysql> create table tbl_2 (col_14 time);
Query OK, 0 rows affected (0.00 sec)

mysql> insert into tbl_2 values (NULL);
Query OK, 1 row affected (0.01 sec)

mysql> alter table tbl_2 change column col_14 col_41 year;
Query OK, 0 rows affected (2.52 sec)
```

File: types/datum.go [hunk 1]
```diff
@@ -1415,8 +1415,6 @@ func (d *Datum) ConvertToMysqlYear(sc *stmtctx.StatementContext, target *FieldTy
 		}
 	case KindMysqlTime:
 		y = int64(d.GetMysqlTime().Year())
-	case KindMysqlDuration:
-		y = int64(time.Now().Year())
```
Comment (tangenta) [inline types/datum.go:1419]: Remove this because of https://github.com/pingcap/tidb/issues/25331#issuecomment-896530427.

**Review by AilinKid — APPROVED** [review]

File: ddl/db_test.go [hunk 2]
```diff
@@ -5173,23 +5173,23 @@ func (s *testSerialDBSuite) TestModifyColumnCharset(c *C) {
...
-		{"time", `0012`, "year", currentYear, 0},
-		{"time", `12`, "year", currentYear, 0},
-		{"time", `"30 20:00:12.498"`, "year", currentYear, 0},
-		{"time", `"20:00:12.498"`, "year", currentYear, 0},
-		{"time", `"200012.498"`, "year", currentYear, 0},
-		{"time", `200012.498`, "year", currentYear, 0},
+		{"time", `"30 20:00:12"`, "year", "", invalidYearCode},
+		{"time", `"30 20:00"`, "year", "", invalidYearCode},
+		{"time", `"30 20"`, "year", "", invalidYearCode},
+		{"time", `"20:00:12"`, "year", "", invalidYearCode},
```
Comment (zimulala) [inline ddl/db_test.go:5182]: This is not the same as MySQL error message：
```
mysql> create table t(a time);
Query OK, 0 rows affected (0.04 sec)

mysql> insert into t(a) values("20:00:12");
Query OK, 1 row affected (0.02 sec)

mysql> alter table t modify a year;
ERROR 1264 (22003): Out of range value for column 'a' at row 1
```

File: ddl/column.go [hunk 3]
```diff
@@ -1388,15 +1388,24 @@ func (w *updateColumnWorker) getRowRecord(handle kv.Handle, recordKey []byte, ra
...
+		err = types.ErrWarnDataOutOfRange.GenWithStack("Out of range value for column '%s', the value is '%s'", w.oldColInfo.Name, dStr)
 	}
 	return err
 }
 
+func datumToStringNoErr(d types.Datum) string {
+	if v, err := d.ToString(); err == nil {
+		return v
+	}
+	return ""
```
Comment (zimulala) [inline ddl/column.go:1406]: Do we need to return `fmt.Sprintf("%v", GetValue())`?

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**